[AArch64][GlobalISel] Fix miscompile on carry-in selection

tobias-stadler opened the pull request:
Eliding the carry-in setting instruction for G_UADDE/... is illegal if it causes the carry generating instruction to become dead because ISel will just remove the instruction.
I accidentally introduced this here: https://reviews.llvm.org/D153164.
As far as I can tell, this is not exposed on the default clang settings, because on O0 there is always a G_AND between boolean defs and uses, so the optimization doesn't apply. Thus, when I tried to commit https://reviews.llvm.org/D159140, which removes these G_ANDs on O0, I broke some UBSan tests.

The optimization is more of a band-aid to cover the most common case, so I'd also be willing to just remove it right now. I'll eventually add some proper NZCV optimizations to cover these cases (and stuff like when the CarryOut is directly fed into a branch) to PostSelectOptimize. Please let me know if we want to keep this as a cheap check for O0 even though it is a little hacky.

tobias-stadler commented:
Would it be legal to just add NZCV as implicit-def to the previous G_UADDE/.. to make the dependency explicit? addOperand would place it at the end of the operand array, so I believe it should be ok.

aemerson commented:
What if we just call `select()` on the now-dead instruction? If that works it seems the most reliable way, if it doesn't then adding NZCV seems like an ok stop-gap.

tobias-stadler committed:
[AArch64][GlobalISel] Fix miscompile on carry-in selection

Eliding the NZCV setting instruction for G_UADDE/... is illegal if it causes the
instruction that defines the carry vReg to become dead. We fix this by recursively
selecting this instruction before continuing to select the current instruction.

aemerson commented on llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp:
```diff
@@ -3552,6 +3557,13 @@ bool AArch64InstructionSelector::select(MachineInstr &I) {
+bool AArch64InstructionSelector::selectAndRestoreState(MachineInstr &I) {
+  MachineIRBuilderState OldMIBState = MIB.getState();
+  bool Success = select(I);
+  MIB.getState() = std::move(OldMIBState);
```
Just add a `setState()` instead? The state can be copied by value, it's just some pointers and an iterator.

tobias-stadler replied:
What about the DebugLoc? It has a TrackingMDRef, which does seem to benefit from moving, because it can call MetadataTracking::retrack(). I don't understand the details of MetadataTracking though.

aemerson replied:
I'm also not familiar with that, but we already copy DebugLocs everywhere in llvm so I doubt think it's an issue?

tobias-stadler committed:
fixup! [AArch64][GlobalISel] Fix miscompile on carry-in selection
